File: scripts/ensure-worker-hidden.py
# File: C:\Users\Yoshi\Documents\GitHub\LINE_AI_AGENT\scripts\ensure-worker-hidden.py
# Summary: Task Schedulerからensure-worker.ps1をコンソール非表示で実行するラッパーです。

from __future__ import annotations

import argparse
import subprocess
import sys
from pathlib import Path


CREATE_NO_WINDOW = 0x08000000
POWERSHELL_EXE = Path(r"C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe")


def parse_args() -> argparse.Namespace:
    """Task Schedulerから渡される監視用引数を解釈します。"""
    script_dir = Path(__file__).resolve().parent
    project_root = script_dir.parent

    parser = argparse.ArgumentParser()
    parser.add_argument("--env-file", default=str(project_root / ".env"))
    parser.add_argument("--log-level", default="INFO")
    return parser.parse_args()


def main() -> int:
    """PowerShell監視スクリプトを子プロセスとして非表示実行し、終了コードを引き継ぎます。"""
    args = parse_args()
    script_dir = Path(__file__).resolve().parent
    project_root = script_dir.parent
    ensure_script = script_dir / "ensure-worker.ps1"

    command = [
        str(POWERSHELL_EXE),
        "-NoProfile",
        "-WindowStyle",
        "Hidden",
        "-ExecutionPolicy",
        "Bypass",
        "-File",
        str(ensure_script),
        "-EnvFile",
        str(Path(args.env_file)),
        "-LogLevel",
        args.log_level,
    ]

    # pythonw.exeには標準コンソールがないため、子PowerShellにもコンソールを割り当てません。
    result = subprocess.run(
        command,
        cwd=str(project_root),
        creationflags=CREATE_NO_WINDOW,
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL,
        check=False,
    )
    return int(result.returncode)


if __name__ == "__main__":
    sys.exit(main())

File: scripts/ensure-worker-hidden.js
// <PROJECT_ROOT>/scripts/ensure-worker-hidden.js
// Windows Task SchedulerからPowerShell監視をコンソールを表示せず実行し、終了コードを返すラッパーです。

(function () {
    if (WScript.Arguments.Length !== 3) {
        WScript.Quit(87);
    }

    function quote(value) {
        // 空白を含むパスもPowerShellの1個の引数として渡せるよう、二重引用符をエスケープします。
        return '"' + String(value).replace(/"/g, '""') + '"';
    }

    var ensureScript = WScript.Arguments.Item(0);
    var envFile = WScript.Arguments.Item(1);
    var logLevel = WScript.Arguments.Item(2);
    var command = "powershell.exe -NoProfile -NonInteractive -ExecutionPolicy Bypass -File " + quote(ensureScript) + " -EnvFile " + quote(envFile) + " -LogLevel " + quote(logLevel);
    var shell = new ActiveXObject("WScript.Shell");

    // 第2引数の0は非表示、第3引数trueは監視スクリプトの終了コードを確実に回収するための待機です。
    WScript.Quit(shell.Run(command, 0, true));
}());
